libraries/mpxe-store/src/store.c

```diff
@@ -0,0 +1,174 @@
+#include "store.h"
+
+#include <fcntl.h>
+#include <poll.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "critical_section.h"
+#include "gpio.pb-c.h"
+#include "log.h"
+
+#define MAX_STORE_COUNT 64
+#define INVALID_STORE_ID MAX_STORE_COUNT
+
+typedef struct Store {
+  MxStoreType type;
+  void *key;
+  void *store;
+} Store;
+
+// Used to ensure this module is only initialized once
+static bool s_initialized = false;
+
+static Store s_stores[MAX_STORE_COUNT];
+
+// Child to parent fifo - used to talk to harness
+static int s_ctop_fifo;
+
+static StoreFuncs s_func_table[MX_STORE_TYPE__END];
+
+static pthread_mutex_t s_sig_lock;
+
+// signal handler for catching parent
+static void prv_sigusr(int signo) {
+  pthread_mutex_unlock(&s_sig_lock);
+  return;
+}
+
+Store *prv_get_first_empty() {
+  for (uint16_t i = 0; i < MAX_STORE_COUNT; i++) {
+    if (s_stores[i].key == NULL && s_stores[i].store == NULL) {
+      return &s_stores[i];
+    }
+  }
+  return NULL;
+}
```

Comment:
Not exactly since objpool literally acts like a heap and doesn't allow keyed storage other than the pointer to the node, whereas in this case I'm trying to use it like a dict. This is sorta mimicing the soft timer implementation so I'm not that fussed, but you're right there's probably a better way to do this.

Reply:
objpool doesn't allow keyed storage of any kind, and doesn't allow iterating. Soft timer implements a similar pattern as this (different cus the softtimerid keys are the array indices), so I'm fairly certain there isn't a significantly easier way to do this